geopandas/tests/test_sindex.py

```diff
@@ -172,10 +181,299 @@ def test_intersection_invalid_bounds_tuple(self, test_geom):
+                "contains",
+                [(-1, -1, 2, 2)],
+                [[0, 0], [0, 1]],
+            ),  # intersects and contains multiple
+        ),
+    )
+    def test_query_bulk(self, predicate, test_geom, expected):
+        """Tests the `query_bulk` method with valid
+        inputs and valid predicates.
+        """
+        # pass through GeoSeries to have GeoPandas
+        # determine if it should use shapely or pygeos geometry objects
+        test_geom = geopandas.GeoSeries(
+            [box(*geom) for geom in test_geom], index=range(len(test_geom))
+        )
+        res = self.df.sindex.query_bulk(test_geom, predicate=predicate)
+        assert_array_equal(res, expected)
+
+    @pytest.mark.parametrize(
+        "test_geoms, expected_value",
+        [
+            # single empty geometry
+            ([GeometryCollection()], [[], []]),
+            # None should be skipped
+            ([GeometryCollection(), None], [[], []]),
+            ([None], [[], []]),
+            ([None, box(-0.5, -0.5, 0.5, 0.5), None], [[1], [0]]),
+        ],
+    )
+    def test_query_bulk_empty_geometry(self, test_geoms, expected_value):
+        """Tests the `query_bulk` method with an empty geometry.
+        """
+        # pass through GeoSeries to have GeoPandas
+        # determine if it should use shapely or pygeos geometry objects
+        # note: for this test, test_geoms (note plural) is a list already
+        test_geoms = geopandas.GeoSeries(test_geoms, index=range(len(test_geoms)))
+        res = self.df.sindex.query_bulk(test_geoms)
+        assert_array_equal(res, expected_value)
+
+    def test_query_bulk_empty_input_array(self):
+        """Tests the `query_bulk` method with an empty input array.
+        """
+        test_array = np.array([], dtype=object)
+        expected_value = [[], []]
+        res = self.df.sindex.query_bulk(test_array)
+        assert_array_equal(res, expected_value)
+
+    def test_query_bulk_invalid_input_geometry(self):
+        """Tests the `query_bulk` method with invalid input for the `geometry` parameter.
+        """
+        test_array = "notanarray"
+        with pytest.raises(TypeError):
+            self.df.sindex.query_bulk(test_array)
+
+    def test_query_bulk_invalid_predicate(self):
+        """Tests the `query_bulk` method with invalid predicates.
+        """
+        test_geom_bounds = (-1, -1, -0.5, -0.5)
+        test_predicate = "test"
+
+        # pass through GeoSeries to have GeoPandas
+        # determine if it should use shapely or pygeos geometry objects
+        test_geom = geopandas.GeoSeries([box(*test_geom_bounds)], index=["0"])
+
+        with pytest.raises(ValueError):
+            self.df.sindex.query_bulk(test_geom.geometry, predicate=test_predicate)
+
+    @pytest.mark.parametrize(
+        "predicate, test_geom, expected",
+        (
+            (None, (-1, -1, -0.5, -0.5), [[], []]),
+            ("intersects", (-1, -1, -0.5, -0.5), [[], []]),
```

Comment:
Since multiple paths through the implementation can produce empty result arrays, it probably would be a good idea to include a case with a non-empty result too.

Reply:
Added a case `("contains", (-1, -1, 1, 1), [[0], [0]])`